### DISCUSSION TITLE
Help Bug or other ?! func.current_timestamp() as a default or an onupdate value emits an unwanted query

### DISCUSSION THREAD

**Discussion by frank2411** [sqlalchemy/sqlalchemy#8877] [Usage Questions] [answered]
Hi,

I don't know if what is happening to me is a bug or I am missing something but I would like some help if possible.

I've got the following base model:

```python

class Model:
    __abstract__ = True

    session: Optional[scoped_session] = None
    query: Optional[Query] = None

    created_at = Column(DateTime(timezone=True), default=func.current_timestamp())
    updated_at = Column(
        DateTime(timezone=True),
        default=func.current_timestamp(),
        onupdate=func.current_timestamp(),
    )

```

Every other model in my code base inherits from this class.

Since at work we loadbalance traffic between master and replicas, we encountered some errors that, based on the hints on the internet, could be resolved setting the `expire_on_commit=False` in our scoped sessions.

This approach, that doesn't (normally) emits new queries after commit on accessing an entity, actually only partially resolved our problems.

In fact, while investigating what was happening, apparently on deserialization (we use marshmallow) after creating or updating an obect, a select query like this is emitted:

```text

2022-11-25 16:22:54,892 INFO sqlalchemy.engine.Engine SELECT users.updated_at AS users_updated_at 
FROM users 
WHERE users.id = %(pk_1)s
2022-11-25 16:22:54,892 INFO sqlalchemy.engine.Engine [cached since 43.28s ago] {'pk_1': 1}
2022-11-25 16:22:54,894 INFO sqlalchemy.engine.Engine ROLLBACK

```

Those are complete transaction logs:

```text

2022-11-25 16:22:54,874 INFO sqlalchemy.engine.Engine SELECT users.created_at, users.updated_at, users.id, users.email, users.password, users.role, users.is_active, users.organization_id, users.last_password_update, users.last_login_date, users.temporary_token, groups_1.created_at AS created_at_1, groups_1.updated_at AS updated_at_1, groups_1.id AS id_1, groups_1.name, groups_1.organization_id AS organization_id_1 
FROM users LEFT OUTER JOIN (users_groups AS users_groups_1 JOIN groups AS groups_1 ON groups_1.id = users_groups_1.group_id) ON users.id = users_groups_1.user_id 
WHERE users.id = %(id_2)s
2022-11-25 16:22:54,874 INFO sqlalchemy.engine.Engine [cached since 43.27s ago] {'id_2': 1}
2022-11-25 16:22:54,879 INFO sqlalchemy.engine.Engine UPDATE users SET updated_at=CURRENT_TIMESTAMP, email=%(email)s WHERE users.id = %(users_id)s
2022-11-25 16:22:54,879 INFO sqlalchemy.engine.Engine [cached since 43.27s ago] {'email': 'cicce@luto.it', 'users_id': 1}
2022-11-25 16:22:54,881 INFO sqlalchemy.engine.Engine COMMIT
2022-11-25 16:22:54,891 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-11-25 16:22:54,892 INFO sqlalchemy.engine.Engine SELECT users.updated_at AS users_updated_at 
FROM users 
WHERE users.id = %(pk_1)s
2022-11-25 16:22:54,892 INFO sqlalchemy.engine.Engine [cached since 43.28s ago] {'pk_1': 1}
2022-11-25 16:22:54,894 INFO sqlalchemy.engine.Engine ROLLBACK


```

Actually the third query should not be executed as per `expire_on_commit=False` and as per what the documentation says about this classic database function ( func.current_timestamp() ).

I resolved the issue putting every as shown:

```python

class Model:
    __abstract__ = True

    session: Optional[scoped_session] = None

    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

```

This way the third unwanted query it's not emitted.

The database is PostgreSQL14.

Can someone explain to me if this is the correct intended behavior ? Or If I should report this as a bug ? 

Or even if I'm missing something.

Thank you very much for your work!

**Comment by zzzeek** [comment, marked as answer]
it looks like the SELECT is getting the value of the server-generated default for created_at.  SQLAlchemy doesnt know what that will be unless it SELECTs it back, or includes it in the RETURNING statement, for backends (like PostgreSQL) that support RETURNING.    Your serializer accesses it, and it has to get loaded.

to include these in RETURNING automatically (PostgreSQL  / SQL Server / Oracle only for 1.4, sqlite / mariadb added in 2.0), use [eager_defaults](https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#orm-server-defaults).

**Reply by frank2411** [reply]
Thank you for the answer! I will check the doc link you sended.

But, still, one question remains: should `expire_on_commit=False` fix this forced reload ?

As per documentation, whit this set to False, Sqlalchemy shouldn't re-trigger object reload.

In fact the strange thing is that I can access every other attribute in the model, like `description` or `email` and correctly no reload is done (correct behavior).

Plus I read the documentation about the `func.current_timestamp` and it seems it should be put as an `inline` in the query. So actually no side query is mentioned.

Did I get it wrong ?

Really appreciate your help!

**Reply by CaselIT** [reply]
Well you specified that `current_timestamp` should be used on update. That's a function that will be executed by the database, so how can sqlalchemy know what value was set by the database without issueing a select or usong returning in the update?
If it did not execute the query the old value (the one preceding the update) would be returned.

In your second example, `datetime.utcnow` is a python function, so nothing is generated on the database side and no select is required after the update to know the new value 

Since you are using postgresql you can specify `eager_default=true` like mike suggested, to make use of `update .. returning` avoiding the select

**Reply by frank2411** [reply]
So, just to be sure I understood the generic concept.

Letting the database generate the value, actually leaves the code "not knowing it" so, in order to correctly populate an updated or created (since I have the same problem on creation) model(or object), another query must be made to retrieve the database generated value.

To avoid this query `eager_default=True` must be used specifiyng `returning ..` like in the doc you provided.

Did I catch the explanation ?

Thanks!

**Reply by CaselIT** [reply]
Yes

Sqlalchemy calls the statis "not knowing it" expired, meaning that it needs to fect the value since the on in python is no longer curent.

Regarding eager_defaults IIRC in v2 that's enabled by default if the dialect has returing

**Reply by zzzeek** [reply]
it is?   I dont think it is.   ORM models dont know about dialects.    it will use returning if the dialect has it, is what it decides.   

we could make "eager_defaults=auto" though and do that too.   this is all because RETURNING is much more widespread now.

**Reply by CaselIT** [reply]
I'm probably wrong now. I thought this was connected to the whole `synchronize_session=auto` that uses returning now.
I guess that the same logic could be applied to eager_defaults if that's a different thing compared to the sync session

**Reply by frank2411** [reply]
Thank you very much! Actually I can confirm that the option `eager_defaults`  it's not activated by default. I'm using sqlalchemy with the 2.0 API (`future=True`) and postgresql supports RETURNING but I have to activate the option per model or the ORM emits a query :)

**Reply by CaselIT** [reply]
@zzzeek do you think it make sense to change it to `auto` and have the same logic as in synchronize_session?

**Reply by zzzeek** [reply]
yes eager_defaults can be "auto", so that we can use a dialect-level check in persistence to know if this is something we get "for free" or not.

**Reply by CaselIT** [reply]
I've created https://github.com/sqlalchemy/sqlalchemy/issues/8889
